### PR TITLE
[exporter/awskinesisexporter] fix in compressor crashing under heady load due non-safe thread execution

### PR DESCRIPTION
**Pull request by leorinat**
Fixing a bug that made the execution panic when the load was high enough, specially if the payloads were not very tiny.

**Testing:** 
Executed this with and without the fix locally using heavy load, then ran it in cloud servers using heavier load and a variety of payloads.

### CODE REVIEW

File: exporter/awskinesisexporter/internal/compress/noop_compression.go [hunk 1]
```diff
@@ -9,12 +9,6 @@ type noop struct {
 	data io.Writer
 }
 
-func NewNoopCompressor() Compressor {
```
Comment (MovieStoreGuy) [inline exporter/awskinesisexporter/internal/compress/noop_compression.go:12]: Is there any reason to remove the method here?

File: .chloggen/fix_compressor-kinesis-exporter-thread-safe.yaml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: the compressor was crashing under high load due it not being thread safe.
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [32589]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext: Instead of using a mutex in the compressor Do I ended up creating a new compressor on each batch generation because it generated more performant execution in my tests with heavy load, specially if the payloads were bigger
```
Comment (MovieStoreGuy) [inline .chloggen/fix_compressor-kinesis-exporter-thread-safe.yaml:18]: This subtext reads more as a first person introspection as opposed to more informative details.

Could I ask you to keep it focused on the improvement?

**Review by MovieStoreGuy — COMMENTED** [review]
Considering that this is intended to resolve a potential race condition, if we can look at ensuring that this can not happen again with tests, or ensure that operations are atomic is the ideal here.
Could you help provide some additional tests, or protections that this will stop happening?

File: exporter/awskinesisexporter/internal/batch/batch.go [hunk 3]
```diff
@@ -78,7 +75,13 @@ func New(opts ...Option) *Batch {
 }
 
 func (b *Batch) AddRecord(raw []byte, key string) error {
-	record, err := b.compression.Do(raw)
+
+	compressor, err := compress.NewCompressor(b.compressionType)
```
Comment (jamesmoessis) [inline exporter/awskinesisexporter/internal/batch/batch.go:79]: I think the source of the race condition is here that `compression.Do` holds the buffer as an instance and so if you call it in parallel, it breaks. 

You've solved this by creating a new compressor each time we want to add a record. I think it's probably better that we just convert these compressors to: 

```
type compressor func(in []byte) []byte
```

So they don't hold any internal state and we don't need to create a new compressor struct every time.

Comment (leorinat) [reply exporter/awskinesisexporter/internal/batch/batch.go:79]: sounds great, I'll do it in this way

File: exporter/awskinesisexporter/internal/compress/compresser.go [hunk 4]
```diff
@@ -9,77 +9,82 @@ import (
 	"compress/gzip"
 	"compress/zlib"
 	"fmt"
-	"io"
 )
 
-type bufferedResetWriter interface {
```
Comment (jamesmoessis) [inline exporter/awskinesisexporter/internal/compress/compresser.go:15]: Nicely done!

**Review by jamesmoessis — APPROVED** [review]
Looks good to me, nice work!

**Comment by jcmackie** [comment]
> Considering that this is intended to resolve a potential race condition, if we can look at ensuring that this can not happen again with tests, or ensure that operations are atomic is the ideal here. Could you help provide some additional tests, or protections that this will stop happening?

It looks like there's a test that might satisfy this requirement here: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32589/files#diff-280135a7ba6930827a3944b3ac3fdc236eed9dc619c336dd817d22c4bd73dc9cR125
Does that look good to you @MovieStoreGuy?

File: exporter/awskinesisexporter/internal/compress/noop_compression.go (hunk 1 above)
Comment (leorinat) [reply exporter/awskinesisexporter/internal/compress/noop_compression.go:12]: the compressor abstraction got removed

**Comment by jcmackie** [comment]
Hello @Aneurysm9!
I think Sean (MovieStoreGuy) is sick at the moment, and I don't want to bother him with a ping.

Would you be able to review this PR for us and merge if you're happy with it?

Regards,
James

**Review by MovieStoreGuy — APPROVED** [review]

**Pull request merged**